### PR TITLE
Add the new `startup.mode.timestamp.start.at.operation.time` config property

### PR DESCRIPTION
**Pull request by stIncMale**
Instead of just introducing a new config property, I observed that the behavior introduced by it is mutually exclusive with the `copy.existing.*` behavior, and both define the connector behavior when there is no offset data. This observation should be reflected in the structure of our source configuration, if we want the configuration to be clear to a user (and us, the maintainers of the connector). So I did the following:

- ~~Introduced the `start` configuration property with two values `ignore_existing` and `copy_existing`. The latter is equivalent to setting `copy.existing = true`.~~
- ~~Deprecated `copy.existing`, which required refactoring the code we use for soft validation of configs. The property is still supported.~~
- ~~Refactored the use of the rest of the `copy.existing.*` properties such that it's clear from the code that they have effect only if `start = copy_existing`.~~
- ~~Introduced and implemented the new `ignore.existing.before.operation.time` property, which has effect only if `start = ignore_existing`.~~

- Introduced the `start.mode` configuration property with three values `latest`, `timestamp`, `copy_existing`. The former is equivalent to the current default behavior, the latter is equivalent to setting `copy.existing = true`.
- The `timestamp` value actuates the new `startup.mode.timestamp.start.at.operation.time` property.
- Deprecated `copy.existing*` properties, which required refactoring the code we use for soft validation of configs. The deprecated properties are still supported.
- Introduced `startup.mode.copy.existing.*` properties as replacements for the deprecated `copy.existing.*` properties.
- Refactored the use of the rest of the `*copy.existing*` properties such that it's clear from the code that they have effect only if `start.mode = copy_existing`.

KAFKA-330

### CODE REVIEW

File: src/main/java/com/mongodb/kafka/connect/sink/SinkConfigSoftValidator.java
Comment (stIncMale) [inline src/main/java/com/mongodb/kafka/connect/sink/SinkConfigSoftValidator.java:68]: This class was split into `ConfigSoftValidator` (contains the validation functionality) and `SinkConfigSoftValidator`/`SourceConfigSoftValidator` that contain specific business rules to check (unused, deprecated, conflicting properties).

File: src/main/java/com/mongodb/kafka/connect/util/config/ConfigSoftValidator.java
Comment (stIncMale) [inline src/main/java/com/mongodb/kafka/connect/util/config/ConfigSoftValidator.java:175]: The `ObsoletePropertiesSet` class is a new piece of code in `ConfigSoftValidator` (the content of `ConfigSoftValidator` was originally in `SinkConfigSoftValidator`).

File: src/main/java/com/mongodb/kafka/connect/util/config/ConfigSoftValidator.java
Comment (stIncMale) [inline src/main/java/com/mongodb/kafka/connect/util/config/ConfigSoftValidator.java:53]: The implementation of the `logObsoleteProperties` has changed from what we originally had in `SinkConfigSoftValidator` before its code was moved here. The `ObsoletePropertiesSet` class is also a new piece of code. The rest of the `ConfigSoftValidator` was just moved here from `SinkConfigSoftValidator`.

**Comment by rozza** [comment]
I think: `start.at.operation.time` would be easier to understand / discover and document than a totally new config `ignore.existing.before.operation.time`. But I get the logic for adding it, so will defer to Robert on config names / deprecating copy.existing.

For the config is there need for the increment or will just seconds since epoch suffice?  Much easier to configure for users if that is the case. Not 100% sure of the use case.  Setting `startAtOperationTime` being to `new BsonTimeStamp(0)` will start at the beginning of the oplog - rather than from the end. Is this something we should default to? 

@RWaltersMA any thoughts?

**Comment by Jiabao-Sun** [comment]
For the config names, how about use `startup.mode` which provides three mode for user to choose.
- full:  `copy.existing = true`
- earliest:  `copy.existing = false` and from the start at the beginning of the oplog
- latest: `copy.existing = false`
- timestamp:  `copy.existing = false` and from the specific timestamp. And an extra option`startup.at.operation.time` is needed for this startup mode.

**Review by Jiabao-Sun — COMMENTED** [review]
https://github.com/mongodb/mongo-kafka/blob/2662926a35411844bd9bcd474ea71a91efcde3d2/src/main/java/com/mongodb/kafka/connect/source/StartedMongoSourceTask.java#L466
The log message `copy.existing=true` may be modified as well.

https://github.com/mongodb/mongo-kafka/blob/2662926a35411844bd9bcd474ea71a91efcde3d2/src/main/java/com/mongodb/kafka/connect/source/StartedMongoSourceTask.java#L440-L457
The `startAtOperationTime` was supported after MongoDB 4.0, we may check this error here rather than return a null ChangeStreamCursor that will dive into an infinite loop.

```
com.mongodb.MongoCommandException: Command failed with error 40415 (Location40415):
 'BSON field '$changeStream.startAtOperationTime' is an unknown field.' on server 
localhost:27017. The full response is {"operationTime": {"$timestamp": {"t": 1667838442, "i": 
1}}, "ok": 0.0, "errmsg": "BSON field '$changeStream.startAtOperationTime' is an unknown 
field.", "code": 40415, "codeName": "Location40415", "$clusterTime": {"clusterTime": 
{"$timestamp": {"t": 1667838442, "i": 1}}, "signature": {"hash": {"$binary": {"base64": 
"AAAAAAAAAAAAAAAAAAAAAAAAAAA=", "subType": "00"}}, "keyId": 0}}}
```

**Comment by stIncMale** [comment]
Thanks, @Jiabao-Sun, [those](https://github.com/mongodb/mongo-kafka/pull/125#pullrequestreview-1170772575) are very helpful catches. Done.

As for your [proposal](https://github.com/mongodb/mongo-kafka/pull/125#issuecomment-1305796075), something like that was my initial approach (though I did not consider exposing something equivalent to `earliest` explicitly as it can be achieved via what is called `timestamp` in your proposal). However, I did not like that `timestamp` has to either require a user to specify `timestamp.start.at.operation.time`<sup>1</sup> (I'd like to avoid such a behavior when possible), or has to explain that it is equivalent to `startup.mode = latest` if `timestamp.start.at.operation.time` is not specified. I can live with that if that is fine by the rest of the team because this gives us the desired `timestamp.start.at.operation.time` property name.

---

<sup>1</sup> By prefixing dependent property names with the `start`/`startup.mode` value they are actuated by, we express the relation via naming, which is why `timestamp.start.at.operation.time` is preferrable to `startup.at.operation.time` (for the same reason the value `full` should rather be `copy_existing`). It would have been even better to additionally include the `start`/`startup.mode` property name in the prefix, but this approach cannot be reconciled with all the `copy.existing.*` properties without deprecating them all, which I anticipate, would have made getting the rest of the team onboard with the approach even harder.

**Comment by rozza** [comment]
Just to note: Kafka has the following option: `auto.offset.reset` which can be `earliest` or `latest` which is the equivalent to setting setting `startAtOperationTime(new BsonTimeStamp(0))` and not setting a `startAtOperationTime` value.

File: src/main/java/com/mongodb/kafka/connect/util/config/BsonTimestampParser.java
Comment (rozza) [inline src/main/java/com/mongodb/kafka/connect/util/config/BsonTimestampParser.java:28]: Like this :)

File: src/main/java/com/mongodb/kafka/connect/util/config/BsonTimestampParser.java
Comment (rozza) [inline src/main/java/com/mongodb/kafka/connect/util/config/BsonTimestampParser.java:43]: Should we log a warning or error if they provide milliseconds eg: `2018-11-30T18:35:24.12Z`?

File: src/main/java/com/mongodb/kafka/connect/util/config/BsonTimestampParser.java
Comment (rozza) [inline src/main/java/com/mongodb/kafka/connect/util/config/BsonTimestampParser.java:34]: Lets add a unit test for this.

**Review by rozza — CHANGES_REQUESTED** [review]

File: src/main/java/com/mongodb/kafka/connect/util/config/BsonTimestampParser.java
Comment (RWaltersMA) [reply src/main/java/com/mongodb/kafka/connect/util/config/BsonTimestampParser.java:43]: I assume ms are not supported ?

**Comment by rozza** [comment]
We can always add documentation in the main docs for how to achieve earliest / latest semantics.

A downside to deprecation is it just generates a log message, so is not very obvious to end users, especially as Kafka connect can generate a lot of log messages.  However, the approach taken still supports these deprecated configs then I'm happy if we can find a cleaner way to express configuration names.

Note suffixing dependant property names should be preferred as; it naturally groups values by the naming convention and is the established convention in Kafka - eg: SMT configuration: https://docs.confluent.io/platform/current/connect/transforms/flatten.html#flatten

File: src/main/java/com/mongodb/kafka/connect/sink/SinkConfigSoftValidator.java
Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/sink/SinkConfigSoftValidator.java:68]: Ack

File: src/main/java/com/mongodb/kafka/connect/source/MongoSourceConfig.java
Comment (rozza) [inline src/main/java/com/mongodb/kafka/connect/source/MongoSourceConfig.java:275]: The first sentence is a little confusing in its relationship to the config.

Suggest simplifying the initial description to something like "The initial connector start up mode."

Then later put a note stating that once the connector has been started and change stream offset data is stored any restarts will continue from that offset.

File: src/main/java/com/mongodb/kafka/connect/source/MongoSourceConfig.java
Comment (rozza) [inline src/main/java/com/mongodb/kafka/connect/source/MongoSourceConfig.java:272]: Suggest: The initial start up behavior when there is no offset data.

File: src/main/java/com/mongodb/kafka/connect/source/StartedMongoSourceTask.java
Comment (rozza) [inline src/main/java/com/mongodb/kafka/connect/source/StartedMongoSourceTask.java:95]: Is this the error for setting startAtOperationTime pre 4.0?

File: src/main/java/com/mongodb/kafka/connect/source/MongoSourceConfig.java
Comment (rozza) [inline src/main/java/com/mongodb/kafka/connect/source/MongoSourceConfig.java:280]: Add note that timestamp requires MongoDB 4.0 or greater.

File: src/main/java/com/mongodb/kafka/connect/source/StartedMongoSourceTask.java
Comment (rozza) [inline src/main/java/com/mongodb/kafka/connect/source/StartedMongoSourceTask.java:459]: This error message is likely to confuse users - this should only happen on MongoDB < 4.0 with a timestamp set. Can you add more information to help users determine the root cause of issue and how to fix?

File: src/main/java/com/mongodb/kafka/connect/util/config/BsonTimestampParser.java
Comment (rozza) [inline src/main/java/com/mongodb/kafka/connect/util/config/BsonTimestampParser.java:37]: This class should have its own stand alone unit tests. Its not obvious that there are existing tests in `MongoSourceConfigTest` - they probably can just be moved over to `BsonTimestampParserTest`.

This way in the future maintainers will find it easier should any changes be required.

**Review by rozza — CHANGES_REQUESTED** [review]
Looking good - couple of nits and suggestions.

**Review by katcharov — APPROVED** [review]
LGTM. I like the approach taken (of having mutually exclusive configs represented under a single property).

File: src/main/java/com/mongodb/kafka/connect/source/MongoSourceConfig.java
Comment (stIncMale) [reply src/main/java/com/mongodb/kafka/connect/source/MongoSourceConfig.java:275]: What does "initial" means here and [below](https://github.com/mongodb/mongo-kafka/pull/125#discussion_r1018830236)? The only meaning I can think of based on your suggestion is "when there is no offset data". If that's the case, then what's the purpose of introducing a new term "initial start up" given that we don't use it anywhere else, including the documentation in question?

> Then later put a note stating that once the connector has been started and change stream offset data is stored any restarts will continue from that offset.

Note that `offset.partition.name` explains nothing about what "offset" means. This seems reasonable given that it's the Connect Framework term, not a connector-specific term.

File: src/main/java/com/mongodb/kafka/connect/source/StartedMongoSourceTask.java
Comment (stIncMale) [reply src/main/java/com/mongodb/kafka/connect/source/StartedMongoSourceTask.java:95]: Yes. But it may also happen with other fields we may introduce in the future, i.e., we should not make the error message specific to `startAtOperationTime`.

**Review by stIncMale — COMMENTED** [review]
For some reason GitHub interprets my comments as part of a review (which I have never started) and shows them as "pending", thus requiring me to add a review.

File: src/main/java/com/mongodb/kafka/connect/util/config/BsonTimestampParser.java
Comment (stIncMale) [reply src/main/java/com/mongodb/kafka/connect/util/config/BsonTimestampParser.java]: Done.

File: src/main/java/com/mongodb/kafka/connect/source/MongoSourceConfig.java
Comment (stIncMale) [reply src/main/java/com/mongodb/kafka/connect/source/MongoSourceConfig.java]: I updated the doc, please take a look.

File: src/main/java/com/mongodb/kafka/connect/source/MongoSourceConfig.java
Comment (stIncMale) [reply src/main/java/com/mongodb/kafka/connect/source/MongoSourceConfig.java:285]: Done.

File: src/main/java/com/mongodb/kafka/connect/source/StartedMongoSourceTask.java
Comment (stIncMale) [reply src/main/java/com/mongodb/kafka/connect/source/StartedMongoSourceTask.java]: Done.

File: src/main/java/com/mongodb/kafka/connect/util/config/BsonTimestampParser.java
Comment (stIncMale) [reply src/main/java/com/mongodb/kafka/connect/util/config/BsonTimestampParser.java:37]: This thread is a duplicate of https://github.com/mongodb/mongo-kafka/pull/125#discussion_r1016544358. Done.

File: src/main/java/com/mongodb/kafka/connect/util/config/BsonTimestampParser.java
Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/util/config/BsonTimestampParser.java]: One is an integration test, the other a unit test. Ideally, you'd only need the unit test for  testing multiple values and one integration test.

File: src/main/java/com/mongodb/kafka/connect/util/config/BsonTimestampParser.java
Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/util/config/BsonTimestampParser.java]: *note: Seconds are supported and there's a test case for that.

**Review by rozza — APPROVED** [review]
LGTM!

File: src/main/java/com/mongodb/kafka/connect/util/config/BsonTimestampParser.java
Comment (stIncMale) [reply src/main/java/com/mongodb/kafka/connect/util/config/BsonTimestampParser.java]: Oh, I wanted to say *nano*seconds :face_palm Thanks for noticing, Ross!

File: src/main/java/com/mongodb/kafka/connect/util/config/BsonTimestampParser.java
Comment (stIncMale) [reply src/main/java/com/mongodb/kafka/connect/util/config/BsonTimestampParser.java]: I agree. Removed some tests from `MongoSourceConfigTest`.

**Pull request merged**
